### PR TITLE
Improve reporting of pipeline with wrong configuration

### LINKED ISSUE #552: Validating "FILE" is displayed too often
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

SOURCES
=======
Validating "UPDATECLI.YAML"

githubSemver
------------
Searching for version matching pattern "~0.1"
✔ Git tag "0.1.2" found matching pattern "~0.1"
Validating "UPDATECLI.YAML"

gitExample2
-----------
ERROR: ✗ no tag found
Validating "UPDATECLI.YAML"

gitExample1
-----------
Searching for version matching pattern "latest"
✔ Git tag "v0.20.1" found matching pattern "latest"
Validating "UPDATECLI.YAML"

githubRegex
-----------
Searching for version matching pattern "v0.*"
✔ Git tag "v0.20.1" found matching pattern "v0.*"

CONDITIONS:
===========
Validating "UPDATECLI.YAML"

### Expected Behavior

In the current state, VALIDATING is displayed each time we re-evaluate a template which is done after each source
I would like to only show VALIDATING once at the beginning of a run

### Steps To Reproduce

_No response_

### Environment

```markdown
- OS: Linux
- updatecli: v0.20.1
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by olblak**
# Improve reporting of pipeline with wrong configuration

Fix #552 
This pull request allows to better identify which pipeline is failing due to a bad configuration. 
Now It would show

```
+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "updatecli/updatecli.d/golang.yaml"
Loading Pipeline "updatecli/updatecli.d/updatecli.yaml"
Loading Pipeline "updatecli/updatecli.d/venom.yaml"
ERROR: failed loading pipeline(s)
	* "updatecli/updatecli.d/golang.yaml" - scms ID "default2" referenced by pullrequest ID "default", doesn't exist
	* "updatecli/updatecli.d/updatecli.yaml" - scm id "default222" doesn't exist

```
instead of the old 
```
+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "updatecli/updatecli.d/golang.yaml"
Validating ""
Validating "GOLANG.YAML"
ERROR: ✗ scms ID "default2" referenced by pullrequest ID "default", doesn't exist
ERROR: Continuing and trying to go as far as possible

```
## Test

To test this pull request, we needs updatecli configuration with wrong content and good content.

```shell
cp <to_package_directory>
go test
```

This pullrequest introduces a behavior change during the "prepare" stage.
Previously , it would just try to load as much configuration as possible and clone as much git repository as possible without failing. But now we exit if we couldn't load any pipeline or if something wrong happened with one of the git clones.

### Potential improvement

Implement the same behavior for git clones where it tries to clones as much repository as possible

### CODE REVIEW

File: pkg/core/engine/main.go [hunk 1]
```diff
@@ -126,43 +127,65 @@ func (e *Engine) Prepare() (err error) {
...
-		if err != nil && err != config.ErrConfigFileTypeNotSupported {
-			logrus.Errorf("%s\n\n", err)
+		if err == config.ErrConfigFileTypeNotSupported {
 			continue
-		} else if err == config.ErrConfigFileTypeNotSupported {
+		} else if err != nil && err != config.ErrConfigFileTypeNotSupported {
+			err = fmt.Errorf("%q - %s", cfgFile, err)
+			errs = append(errs, err)
 			continue
 		}
```
Comment (dduportal) [inline pkg/core/engine/main.go:163]: this if/else struct starts to be hard to read. Do you mind using a switch case there?

Comment (olblak) [reply pkg/core/engine/main.go:163]: Sure, I definitely can

File: pkg/core/config/main.go [hunk 2]
```diff
@@ -131,8 +131,6 @@ func (config *Config) Display() error {
 // Validate run various validation test on the configuration and update fields if necessary
 func (config *Config) Validate() error {
 
-	logrus.Infof("Validating %q", config.Name)
```
[the remark this answers is not in the document]
Comment (lemeurherve) [reply pkg/core/config/main.go:134]: https://github.com/updatecli/updatecli/issues/536

File: pkg/core/engine/main.go (hunk 1 above)
Comment (lemeurherve) [reply pkg/core/engine/main.go:163]: Can't it be just to "if" without "else" here?

One "if err == config.ErrConfigFileTypeNotSupported {"
Then one "if err != nil && err != config.ErrConfigFileTypeNotSupported {"

File: pkg/core/engine/main.go [hunk 3]
```diff
@@ -126,43 +127,65 @@ func (e *Engine) Prepare() (err error) {
...
+			err := fmt.Errorf("%q - %s", cfgFile, err)
+			errs = append(errs, err)
 		}
 
-		e.Pipelines = append(e.Pipelines, newPipeline)
-		e.configurations = append(e.configurations, loadedConfiguration)
+	}
+
+	if len(errs) > 0 {
+
```
Comment (lemeurherve) [inline pkg/core/engine/main.go:182]:
```suggestion
```

File: pkg/core/engine/main.go (hunk 1 above)
Comment (olblak) [reply pkg/core/engine/main.go:163]: I would also prefer switch case as it makes things more readable

Comment (olblak) [reply pkg/core/engine/main.go]: I think it's nitpicking but I took the opportunity to a comment

File: pkg/core/engine/main.go [hunk 4]
```diff
@@ -126,43 +127,67 @@ func (e *Engine) Prepare() (err error) {
...
 	for _, cfgFile := range GetFiles(e.Options.File) {
 
 		loadedConfiguration, err := config.New(cfgFile, e.Options.ValuesFiles, e.Options.SecretsFiles)
 
-		if err != nil && err != config.ErrConfigFileTypeNotSupported {
-			logrus.Errorf("%s\n\n", err)
+		switch {
+		case err == config.ErrConfigFileTypeNotSupported:
+			// Updatecli accepts either a single configuration file or a directory containing multiple configurations.
+			// When browsing files from a directory, we don't want to record error due to unsupported files.
```
Comment (dduportal) [inline pkg/core/engine/main.go:160]:
```suggestion
		switch err.(type) {
		case config.ErrConfigFileTypeNotSupported:
			// Updatecli accepts either a single configuration file or a directory containing multiple configurations.
			// When browsing files from a directory, we don't want to record error due to unsupported files.
```

File: pkg/core/engine/main.go [hunk 5]
```diff
@@ -126,43 +127,67 @@ func (e *Engine) Prepare() (err error) {
...
 			continue
-		} else if err == config.ErrConfigFileTypeNotSupported {
+		case err != nil && err != config.ErrConfigFileTypeNotSupported:
```
Comment (dduportal) [inline pkg/core/engine/main.go:162]:
```suggestion
		case config.ErrConfigFileTypeNotSupported:
```

**Review by dduportal — DISMISSED** [review]
Fine for me, but proposed improvement around the switch by asserting on the error's type.

It's not blocking but as soon as there is an `else` it's a smell: thanks for commenting as it helps to maintain this code (unless there is a unit test to help other maintainers than the one whom wrote the initial conditions).

File: pkg/core/engine/main.go (hunk 5 above)
Comment (lemeurherve) [reply pkg/core/engine/main.go:162]: I didn't realized the "nil" case you were speaking about, it's clear now, thx!

Comment (dduportal) [reply pkg/core/engine/main.go:162]: Yeah, it was there for a (really) good reason. I personnaly like the switch with type assertion: easier to read.

File: pkg/core/engine/main.go (hunk 4 above)
Comment (olblak) [reply pkg/core/engine/main.go]: This suggestion won't work as `config.ErrConfigFileTypeNotSupported` is not a type

File: pkg/core/engine/main.go (hunk 5 above)
Comment (olblak) [reply pkg/core/engine/main.go]: @dduportal I implemented your suggestion

**Review by dduportal — APPROVED** [review]

**Pull request merged**